### PR TITLE
Edit profile page

### LINKED ISSUE #108: Adding a profile edit menu
Related to #42.

### LINKED ISSUE #109: Creating proper navigation and integration with the differentes pages (leaderboard, settings, edit)
Related to #42.

### LINKED ISSUE #110: Adding profile pictures with Firebase
Related to #42.

### PR DESCRIPTION
**Pull request by Maeeen**
All in the title.

This PR adds:
* A profile edit page!
* LiveLazyRefs, LazyRefs that can be listened for real-time changes :+1:

https://user-images.githubusercontent.com/32419465/231849527-3a8fe1b3-967c-4676-8ba3-7daf8b5017d4.mp4

This PR closes #108, closes #109 and closes #110.

### CODE REVIEW

**Comment by BoyeGuillaume** [comment]
> All in the title

Loving it man !!!

**Comment by Maeeen** [comment]
> > All in the title
> 
> Loving it man !!!

If you loved it, then accept it 😤

**Review by BoyeGuillaume — CHANGES_REQUESTED** [review]
Good PR overall, however the backend system could use some improvement. The biggest one is to transfer of all functionalities of `LiveLazyRef` to `LazyRef` so that we could use it everywhere. Ultimately we keep reference to the database **in order** to have live flow of data. Furthermore can keep the `BaseLazyRef` to only handle static stale data while adding `LiveLazyRef` to enable dynamically tracked data. Finally to enable the UI to benefit from this modification (aka. to be re-compose on a need-to basis) you only have to modify `RememberLazyRef` api

File: app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt:19]: > The idea of having live state in LazyRef was a bit the reason why I created this class in the first place.

This was not written anywhere so I could not know :shrug:. I thought the main reason was to actually hold a "reference" to a database element (as you mentionned).

> Ultimately if we have LazyRef everywhere is because we want to have a system of "updates" and "listeners". On that note my own branch would beneficiate such a system.

Then, every lazy ref should implement live lazy refs? This is not a "good" idea. I wish I could use mixins…

> On that note my own branch would beneficiate such a system.

OK.

> Because normally no one has used LazyRef directly within a UI context you should only have to update (along with their instantiation) the RememberLazyRef API.

No, adding a disposable effect to every component that remembers a lazy ref is not a good idea: creating too many listeners for every components is too much. It should be used only when necessary.

File: app/src/main/java/com/github/geohunt/app/ui/RememberLazyRef.kt
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/ui/RememberLazyRef.kt:45]: As said above, I disagree xd

File: firebase-debug.log
Comment (Maeeen) [reply firebase-debug.log:1]: It was yours. It'd be removed when merged? I will remove it manually ofc.

File: app/src/main/java/com/github/geohunt/app/model/database/api/User.kt
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/model/database/api/User.kt:35]: > this would be straitforward.

No it would not. If you actually check `FirebaseBitmapRef.kt`, you will see that every download is cached. I can not force to redownload a image without doing HUGE modifications that will affect everyone's codebase.

File: firebase-debug.log
Comment (BoyeGuillaume) [reply firebase-debug.log:1]: Yeah sorry for the mistake then

File: app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt:19]: Thinking back to what you said, I do not think that every lazy ref should be live. For instance, a challenge's image.

Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt:19]: And again, having a Live lazy ref requires a specific implementation (that could be too much for all the existing lazy refs).

Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/model/LiveLazyRef.kt:19]: Update?

**Review by dhalilov — APPROVED** [review]
Great code ! As the PR is huge, I only skimmed through the tests, which looked correct to me. We can merge this once @BoyeGuillaume approves.

**Review by BoyeGuillaume — APPROVED** [review]
We can fix this issue/refactory LazyRef next sprint

**Review by BoyeGuillaume — DISMISSED** [review]

**Review by BoyeGuillaume — DISMISSED** [review]

**Review by BoyeGuillaume — APPROVED** [review]

**Pull request merged**
